File: backend/internal/handler/audit.go
package handler

import (
	"net/http"

	"github.com/yunuskargi/confbox/internal/auth"
	"github.com/yunuskargi/confbox/internal/database"
	"github.com/yunuskargi/confbox/internal/models"
)

func ListAuditLogs(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	if user.Role != "admin" {
		writeError(w, 403, "Admin only")
		return
	}

	action := queryStr(r, "action")
	resourceType := queryStr(r, "resource_type")
	username := queryStr(r, "username")
	limit := queryInt(r, "limit", 50)
	offset := queryInt(r, "offset", 0)

	if limit > 500 {
		limit = 500
	}

	query := "SELECT id, username, action, resource_type, resource_name, detail, ip_address, created_at FROM audit_logs WHERE 1=1"
	args := []any{}

	if action != "" {
		query += " AND action = ?"
		args = append(args, action)
	}
	if resourceType != "" {
		query += " AND resource_type = ?"
		args = append(args, resourceType)
	}
	if username != "" {
		query += " AND username = ?"
		args = append(args, username)
	}

	query += " ORDER BY created_at DESC LIMIT ? OFFSET ?"
	args = append(args, limit, offset)

	rows, err := database.DB.Query(query, args...)
	if err != nil {
		writeError(w, 500, "Database error")
		return
	}
	defer rows.Close()

	var result []models.AuditLogOut
	for rows.Next() {
		var log models.AuditLogOut
		rows.Scan(&log.ID, &log.Username, &log.Action, &log.ResourceType, &log.ResourceName, &log.Detail, &log.IPAddress, &log.CreatedAt)
		result = append(result, log)
	}
	if result == nil {
		result = []models.AuditLogOut{}
	}
	writeJSON(w, 200, result)
}

File: backend/app/routers/audit.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import AuditLog, User
from app.schemas import AuditLogOut
from app.auth import get_current_user

router = APIRouter(prefix="/api/audit", tags=["audit"])


@router.get("", response_model=list[AuditLogOut])
def list_audit_logs(
    action: Optional[str] = None,
    resource_type: Optional[str] = None,
    username: Optional[str] = None,
    limit: int = Query(50, le=500),
    offset: int = 0,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    if user.role != "admin":
        from fastapi import HTTPException
        raise HTTPException(status_code=403, detail="Admin only")
    q = db.query(AuditLog)
    if action:
        q = q.filter(AuditLog.action == action)
    if resource_type:
        q = q.filter(AuditLog.resource_type == resource_type)
    if username:
        q = q.filter(AuditLog.username == username)
    return q.order_by(AuditLog.created_at.desc()).offset(offset).limit(limit).all()
